### PR TITLE
tls/x509utils: IsSelfSigned() and other validation helpers

### PR DESCRIPTION
**Pull request by amery**
* IsSelfSigned()
* PrivateKeyEqual()
* PublicKeyEqual()
* ValidKeyPair()

## Summary by CodeRabbit

- **New Features**
	- Introduced functions to compare private and public keys for equivalence.
	- Added a validation function for checking if a public key matches a private key.
- **Bug Fixes**
	- Updated the key validation logic to enhance accuracy in certificate and key pair checks.
	- Improved checks for self-signed certificates by refining comparison logic.

### CODE REVIEW

File: tls/x509utils/pair.go [hunk 1]
```diff
@@ -1,18 +1,66 @@
...
+}
+
+// PublicKeyEqual tells if two private keys are the same.
+// nil keys aren't considered comparable.
+func PublicKeyEqual(a, b crypto.PublicKey) bool {
+	if a2, ok := a.(PublicKey); ok {
+		return a2.Equal(b)
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/x509utils/pair.go:25]: `PublicKey` and `PrivateKey` interfaces are defined in `types.go`. same package.

File: tls/x509utils/pair.go [hunk 2]
```diff
@@ -1,18 +1,66 @@
...
+)
+
+// PrivateKeyEqual tells if two private keys are the same.
+// nil keys aren't considered comparable.
+func PrivateKeyEqual(a, b crypto.PrivateKey) bool {
+	if a2, ok := a.(PrivateKey); ok {
+		return a2.Equal(b)
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/x509utils/pair.go:16]: defined in `types.go`

File: tls/x509utils/pair.go [hunk 3]
```diff
@@ -1,18 +1,66 @@
...
 	}
-	pub, ok := cert.PublicKey.(PublicKey)
-	if !ok {
-		// invalid - unreachable
+
+	if pk, ok := key.(PrivateKey); ok {
+		return PublicKeyEqual(cert.PublicKey, pk.Public())
+	}
+
+	return false
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/x509utils/pair.go:38]: defined in `types.go`

File: tls/x509utils/pair.go [hunk 4]
```diff
@@ -1,18 +1,66 @@
...
+}
+
+// ValidKeyPair confirms the public key matches the private one.
+// nil keys aren't considered comparable.
+func ValidKeyPair(pub crypto.PublicKey, key crypto.PrivateKey) bool {
```
Comment (karasz) [inline tls/x509utils/pair.go:43]: `func MatchingKeys(pub crypto.PublicKey, key crypto.PrivateKey) bool {` ?

Comment (amery) [reply tls/x509utils/pair.go:43]: > `func MatchingKeys(pub crypto.PublicKey, key crypto.PrivateKey) bool {` ?

why do you think this doesn't fit the `ValidFoo(...) bool` pattern?

Comment (karasz) [reply tls/x509utils/pair.go:43]: In my mind `Valid` certifies that something is correct, valid, flawless. `Match` means that it is like something else, similar to it, indistinguishable from it. This is more a philosophic observation. If a public key matches the private one does it make it valid? Or is it just computable from the private one?

Comment (amery) [reply tls/x509utils/pair.go:43]: > In my mind `Valid` certifies that something is correct, valid, flawless. `Match` means that it is like something else, similar to it, indistinguishable from it. This is more a philosophic observation. If a public key matches the private one does it make it valid? Or is it just computable from the private one?

I think that next step falls under the `Verify(....) error` pattern

Comment (amery) [reply tls/x509utils/pair.go:43]: I mean, we are using a weaker test on `ValidCertKeyPair()`. Only comparing public keys match, not even the `notBefore`/`notAfter`.

Comment (karasz) [reply tls/x509utils/pair.go:43]: On second thoughts `ValidKeyPair` is ok, if you read it carefully it is only saying that the `KeyPair` is valid, as in it is a pair.

**Review by karasz — APPROVED** [review]
LGTM

**Pull request merged**